src/users/amoc_users_sup.erl

```diff
@@ -1,60 +1,167 @@
+start_link() ->
+    supervisor:start_link({local, ?MODULE}, ?MODULE, no_args).
+
+%% @private
+-spec init(no_args) -> {ok, {supervisor:sup_flags(), [supervisor:child_spec()]}}.
+init(no_args) ->
+    Specs = [
+             #{
+               id => {amoc_users_worker_sup, N},
+               start => {amoc_users_worker_sup, start_link, [N]},
+               restart => permanent,
+               shutdown => infinity,
                type => worker,
-               modules => [amoc_user]},
-    {ok, {SupFlags, [AChild]}}.
+               modules => [amoc_users_worker_sup]
+              }
+             || N <- lists:seq(1, erlang:system_info(schedulers_online)) ],
+    Strategy = #{strategy => one_for_one, intensity => 0},
+    {ok, {Strategy, Specs}}.
+
+%% API
+-spec init_storage() -> any().
+init_storage() ->
+    UserSups = lists:reverse(supervisor:which_children(?MODULE)),
+    UserSupPids = [ Pid || {_, Pid, _, _} <- UserSups ],
+    UserSupPidsTuple = erlang:list_to_tuple(UserSupPids),
+    NumOfSupervisors = tuple_size(UserSupPidsTuple),
+    Atomics = atomics:new(1 + NumOfSupervisors, [{signed, false}]),
+    Storage = #storage{user_count = Atomics, sups = UserSupPidsTuple},
+    persistent_term:put(?MODULE, Storage).
+
+-spec clean_storage() -> any().
+clean_storage() ->
+    persistent_term:erase(?MODULE).
+
+-spec count_no_of_users() -> count().
+count_no_of_users() ->
+    #storage{user_count = Atomics} = persistent_term:get(?MODULE),
+    atomics:get(Atomics, 1).
+
+-spec incr_no_of_users(non_neg_integer()) -> any().
+incr_no_of_users(SupNum) ->
+    #storage{user_count = Atomics} = persistent_term:get(?MODULE),
+    atomics:add(Atomics, 1, 1),
+    atomics:add(Atomics, SupNum + 1, 1).
+
+-spec decr_no_of_users(non_neg_integer()) -> ok.
+decr_no_of_users(SupNum) ->
+    #storage{user_count = Atomics} = persistent_term:get(?MODULE),
+    atomics:sub(Atomics, SupNum + 1, 1),
+    case atomics:sub_get(Atomics, 1, 1) of
+        0 ->
+            amoc_controller:zero_users_running();
+        _ ->
+            ok
+    end.
+
+-spec start_child(amoc:scenario(), amoc_scenario:user_id(), any()) -> ok.
+start_child(Scenario, Id, ScenarioState) ->
+    Sup = get_sup_for_user_id(Id),
+    gen_server:cast(Sup, {start_child, Scenario, Id, ScenarioState}).
+
+%% Group all children based on ID to their respective worker supervisor and cast a request with each
+%% group at once. This way we reduce the number of casts to each worker to always one, instead of
+%% depending on the number of users.
+-spec start_children(amoc:scenario(), [amoc_scenario:user_id()], any()) -> ok.
+start_children(Scenario, UserIds, ScenarioState) ->
+    KeyFun = fun(UserId) ->
+                     get_sup_for_user_id(UserId)
+             end,
+    Assignments = maps:groups_from_list(KeyFun, UserIds),
+    CastFun = fun (Sup, Users) ->
+                      gen_server:cast(Sup, {start_children, Scenario, Users, ScenarioState})
```

Comment:
never expose details of internal `gen_server` protocol, create interface at `amoc_users_worker_sup` module and call it instead.